File: aircon-converters/gson/src/main/java/com/ironsource/aura/aircon/converter/gson/GsonConverter.java
package com.ironsource.aura.aircon.converter.gson;

import com.google.gson.Gson;
import com.google.gson.JsonSyntaxException;
import com.ironsource.aura.aircon.JsonConverter;

import org.jetbrains.annotations.NotNull;

import java.lang.reflect.Type;

public class GsonConverter
		implements JsonConverter {

	private Gson mGson;

	public GsonConverter() {
		this(new Gson());
	}

	public GsonConverter(final Gson gson) {
		mGson = gson;
	}

	@NotNull
	@Override
	public String toJson(@NotNull final Object obj) {
		return mGson.toJson(obj);
	}

	@Override
	public <T> T fromJson(@NotNull final String json, @NotNull final Class<T> clazz) throws
			JsonException {
		try {
			return mGson.fromJson(json, clazz);
		} catch (JsonSyntaxException e) {
			throw new JsonException(e);
		}
	}

	@Override
	public <T> T fromJson(@NotNull final String json, @NotNull final Type type) throws
			JsonException {
		try {
			return mGson.fromJson(json, type);
		} catch (JsonSyntaxException e) {
			throw new JsonException(e);
		}
	}
}

File: aircon-converters/gson/src/main/java/com/ironsource/aura/airconkt/converter/gson/GsonConverter.kt
package com.ironsource.aura.airconkt.converter.gson

import com.google.gson.Gson
import com.google.gson.JsonSyntaxException
import com.ironsource.aura.airconkt.JsonConverter
import com.ironsource.aura.airconkt.JsonException
import com.ironsource.aura.airconkt.utils.Fail
import com.ironsource.aura.airconkt.utils.Response
import com.ironsource.aura.airconkt.utils.Success
import java.lang.reflect.Type

class GsonConverter constructor(private val gson: Gson = Gson()) : JsonConverter {

    override fun toJson(obj: Any?): String? {
        return gson.toJson(obj)
    }

    override fun <T> fromJson(json: String?, clazz: Class<T>): Response<T, JsonException> {
        return try {
            Success(gson.fromJson(json, clazz))
        } catch (e: JsonSyntaxException) {
            Fail(JsonException(e))
        }
    }

    override fun <T> fromJson(json: String?, type: Type): Response<T, JsonException> {
        return try {
            Success(gson.fromJson(json, type))
        } catch (e: JsonSyntaxException) {
            Fail(JsonException(e))
        }
    }
}